### PR TITLE
Bug fix for GSL precip type

### LINKED ISSUE #860: Bug fix for GSL precip type for RRFS
GSL colleagues noted spurious categorical freezing precip in the Arctic in RRFS-A.  This bug fix addresses this issue within the GSL precip type diagnosis.

### PR DESCRIPTION
**Pull request by EricJames-NOAA**
GSL colleagues noted spurious categorical freezing rain in the Arctic in RRFS-A.  This bug fix solves the problem, by using a consistent snow ratio "SR" throughout the GSL precip type diagnosis, directly read from netCDF files.  

The code change was tested for RRFS-A using a standalone UPP test on Jet.

### CODE REVIEW

**Comment by WenMeng-NOAA** [comment]
@EricJames-NOAA Can you sync your branch with the latest UPP develop?

**Comment by WenMeng-NOAA** [comment]
@EricJames-NOAA Do you have the sample model history files for my UPP standalone test?

File: sorc/ncep_post.fd/SURFCE.f [hunk 1]
```diff
@@ -5168,9 +5168,9 @@ SUBROUTINE SURFCE
 !  ---------------------------------------------------------------
 !      Snow-to-total ratio to be used below
 !  ---------------------------------------------------------------
-               snowratio = snow_bucket(i,j)*1.e-3 / (totprcp-graup_bucket(i,j)*1.e-3)
+!              snowratio = snow_bucket(i,j)*1.e-3 / (totprcp-graup_bucket(i,j)*1.e-3)
 
-!              snowratio = SR(i,j)
+               snowratio = SR(i,j)
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/SURFCE.f]: @EricJames-NOAA Can you add a change log SURFCE.f?

**Comment by EricJames-NOAA** [comment]
> @EricJames-NOAA Do you have the sample model history files for my UPP standalone test?

@WenMeng-NOAA my colleague Ming Hu has some RRFS-A history files here on Hera:
/scratch1/BMC/wrfruc/mhu/temp

**Comment by WenMeng-NOAA** [comment]
> > @EricJames-NOAA Do you have the sample model history files for my UPP standalone test?
> 
> @WenMeng-NOAA my colleague Ming Hu has some RRFS-A history files here on Hera: /scratch1/BMC/wrfruc/mhu/temp

@EricJames-NOAA Please see my test results at /scratch1/NCEPDEV/stmp2/Wen.Meng/post_rrfs_2024011718 on Hera, let me know if you see issues. Thanks!

**Comment by WenMeng-NOAA** [comment]
The UPP RTs were conducted on Hera, here are changed results:
'''
hrrr -- WRFTWO
93:63154001:CSNOW:surface:rpn_corr=0.697061:rpn_rms=0.00476052
95:63154727:CFRZR:surface:rpn_corr=0.895791:rpn_rms=0.00195222
96:63155101:CRAIN:surface:rpn_corr=0.999859:rpn_rms=0.00404401

rap -- WRFPRS
674:183977957:CSNOW:surface:rpn_corr=0.975621:rpn_rms=0.0209548
676:183982455:CFRZR:surface:rpn_corr=0.987641:rpn_rms=0.0179119
677:183992462:CRAIN:surface:rpn_corr=0.999638:rpn_rms=0.0126407

rap -- WRFNAT
1049:274956580:CSNOW:surface:rpn_corr=0.975621:rpn_rms=0.0209548
1051:274961078:CFRZR:surface:rpn_corr=0.987641:rpn_rms=0.0179119
1052:274971085:CRAIN:surface:rpn_corr=0.999638:rpn_rms=0.0126407

rrfs (fv3r) -- NATLEV
1381:1083289546:CSNOW:surface:rpn_corr=0.378282:rpn_rms=0.00295225
1383:1083289981:CFRZR:surface:rpn_corr=0.333201:rpn_rms=0.00278381
1384:1083290281:CRAIN:surface:rpn_corr=0.999998:rpn_rms=0.000695186

rrfs(fv3r) -- PRSLEV
845:810214560:CSNOW:surface:rpn_corr=0.378282:rpn_rms=0.00295225
847:810214995:CFRZR:surface:rpn_corr=0.333201:rpn_rms=0.00278381
848:810215295:CRAIN:surface:rpn_corr=0.999998:rpn_rms=0.000695186
'''
@EricJames-NOAA I would think the above changes in GSL precip types are expected. Please let me know if you have comments.

**Comment by EricJames-NOAA** [comment]
> /scratch1/NCEPDEV/stmp2/Wen.Meng/post_rrfs_2024011718

@WenMeng-NOAA I see lots of differences between your results and my results...here are the differences for the p-type variables (left is your results, right is my results):

< 840:4752446371:vt=2024011719:surface:1 hour fcst:CSNOW Categorical Snow [-]:
<     ndata=14452641:undef=0:mean=0.354192:min=0:max=1
> 845:5138042999:vt=2024011719:surface:1 hour fcst:CSNOW Categorical Snow [-]:
>     ndata=14452641:undef=0:mean=0.0641159:min=0:max=1

< 841:4752527246:vt=2024011719:surface:1 hour fcst:CICEP Categorical Ice Pellets [-]:
<     ndata=14452641:undef=0:mean=0.354193:min=0:max=1
> 846:5138304569:vt=2024011719:surface:1 hour fcst:CICEP Categorical Ice Pellets [-]:
>     ndata=14452641:undef=0:mean=0.000986187:min=0:max=1

< 842:4752608120:vt=2024011719:surface:1 hour fcst:CFRZR Categorical Freezing Rain [-]:
<     ndata=14452641:undef=0:mean=0:min=0:max=0
> 847:5138314535:vt=2024011719:surface:1 hour fcst:CFRZR Categorical Freezing Rain [-]:
>     ndata=14452641:undef=0:mean=0.0023521:min=0:max=1

< 843:4752608311:vt=2024011719:surface:1 hour fcst:CRAIN Categorical Rain [-]:
<     ndata=14452641:undef=0:mean=0.645807:min=0:max=1
> 848:5138347893:vt=2024011719:surface:1 hour fcst:CRAIN Categorical Rain [-]:
>     ndata=14452641:undef=0:mean=0.0766009:min=0:max=1

**Comment by WenMeng-NOAA** [comment]
@EricJames-NOAA It seems variable order from your test and my test are different.  You may access the job card of my test /scratch1/NCEPDEV/stmp2/Wen.Meng/eric/UPP/submit_run_post_rrfs_hera.sh. I saved my test to /scratch1/NCEPDEV/stmp2/Wen.Meng/eric/post_rrfs_2024011718. Please let me know if you find any inconsistent configuration.

**Comment by EricJames-NOAA** [comment]
> /scratch1/NCEPDEV/stmp2/Wen.Meng/eric/UPP/submit_run_post_rrfs_hera.sh.

@WenMeng-NOAA your results are now consistent with mine!  And your results from the UPP RTs seem reasonable.  If needed, I can put a flag to only use consistent snow ratio for model name FV3R, and preserve previous code for RAP/HRRR.  Let me know what you think.

**Comment by WenMeng-NOAA** [comment]
> > /scratch1/NCEPDEV/stmp2/Wen.Meng/eric/UPP/submit_run_post_rrfs_hera.sh.
> 
> @WenMeng-NOAA your results are now consistent with mine! And your results from the UPP RTs seem reasonable. If needed, I can put a flag to only use consistent snow ratio for model name FV3R, and preserve previous code for RAP/HRRR. Let me know what you think.

@EricJames-NOAA It would be good that a flag for preserve the configuration for ops. hrrr/rap. Thanks!

**Comment by EricJames-NOAA** [comment]
> > > /scratch1/NCEPDEV/stmp2/Wen.Meng/eric/UPP/submit_run_post_rrfs_hera.sh.
> > 
> > 
> > @WenMeng-NOAA your results are now consistent with mine! And your results from the UPP RTs seem reasonable. If needed, I can put a flag to only use consistent snow ratio for model name FV3R, and preserve previous code for RAP/HRRR. Let me know what you think.
> 
> @EricJames-NOAA It would be good that a flag for preserve the configuration for ops. hrrr/rap. Thanks!

@WenMeng-NOAA the revised code is ready for your testing.  I get identical results for my RRFS test.

**Comment by WenMeng-NOAA** [comment]
@EricJames-NOAA Thanks for updates. We will wrap your PR for final UPP RT.

**Comment by WenMeng-NOAA** [comment]
@FernandoAndrade-NOAA You may start the RT on Hera and Orion. 
For your reference, my UPP RT was completed on Hera with baseline changes in fv3r as:
```
-rw-r--r-- 1 Wen.Meng stmp    75 Jan 24 23:23 rap_2020072316_pe_test/WRFPRS.GrbF16.diff
-rw-r--r-- 1 Wen.Meng stmp  2027 Jan 24 23:25 fv3r_2023062800/PRSLEV10.tm00.diff
-rw-r--r-- 1 Wen.Meng stmp  2027 Jan 24 23:25 fv3r_2023062800_pe_test/PRSLEV10.tm00.diff
-rw-r--r-- 1 Wen.Meng stmp 14603 Jan 24 23:27 fv3r_2023062800/NATLEV10.tm00.diff
-rw-r--r-- 1 Wen.Meng stmp 14603 Jan 24 23:27 fv3r_2023062800_pe_test/NATLEV10.tm00.diff
```

The actual field value changes are in
```
845:810214560:CSNOW:surface:rpn_corr=0.378282:rpn_rms=0.00295225
847:810214995:CFRZR:surface:rpn_corr=0.333201:rpn_rms=0.00278381
848:810215295:CRAIN:surface:rpn_corr=0.999998:rpn_rms=0.000695186
```

**Comment by FernandoAndrade-NOAA** [comment]
> @FernandoAndrade-NOAA You may start the RT on Hera and Orion. For your reference, my UPP RT was completed on Hera with baseline changes in fv3r as:
> 
> ```
> -rw-r--r-- 1 Wen.Meng stmp    75 Jan 24 23:23 rap_2020072316_pe_test/WRFPRS.GrbF16.diff
> -rw-r--r-- 1 Wen.Meng stmp  2027 Jan 24 23:25 fv3r_2023062800/PRSLEV10.tm00.diff
> -rw-r--r-- 1 Wen.Meng stmp  2027 Jan 24 23:25 fv3r_2023062800_pe_test/PRSLEV10.tm00.diff
> -rw-r--r-- 1 Wen.Meng stmp 14603 Jan 24 23:27 fv3r_2023062800/NATLEV10.tm00.diff
> -rw-r--r-- 1 Wen.Meng stmp 14603 Jan 24 23:27 fv3r_2023062800_pe_test/NATLEV10.tm00.diff
> ```
> 
> The actual field value changes are in
> 
> ```
> 845:810214560:CSNOW:surface:rpn_corr=0.378282:rpn_rms=0.00295225
> 847:810214995:CFRZR:surface:rpn_corr=0.333201:rpn_rms=0.00278381
> 848:810215295:CRAIN:surface:rpn_corr=0.999998:rpn_rms=0.000695186
> ```

Just leaving an update here, Hera is still running and Orion has run into some sort of connection issues during the sbatch steps, I'll retry Orion and check on Hera results tomorrow to confirm consistent changes. Thanks!

**Comment by WenMeng-NOAA** [comment]
The UPP RTs were completed on WCOSS2 and Hera. Baseline recreation for fv3r is needed with this PR.

**Comment by FernandoAndrade-NOAA** [comment]
@WenMeng-NOAA my Hera results are consistent with yours with changes in 3 field values within the fv3r case for NATLEV and PRSLEV. Orion had a SLURM configuration issue after maintenance yesterday but it looks fine now and is currently running RTs.

I should also note a cmake error in the Hera log about an incorrect older version being loaded. It did not cause any issues in building or results, and may have been due to me running another WM test, however if it reappears in future testing it would be a good idea to start tracking as a repo issue. I'll keep an eye out for it.

**Comment by FernandoAndrade-NOAA** [comment]
Orion test results match with the Hera tests, with the expected changes in fv3r. Everything looks good. Baselines will be updated for fv3r on Hera and Orion.

In regards to the Hera cmake version error, it looks like it's overwritten with the correct 3.23.1 immediately following the error.

**Review by FernandoAndrade-NOAA — APPROVED** [review]

**Comment by WenMeng-NOAA** [comment]
> @WenMeng-NOAA my Hera results are consistent with yours with changes in 3 field values within the fv3r case for NATLEV and PRSLEV. Orion had a SLURM configuration issue after maintenance yesterday but it looks fine now and is currently running RTs.
> 
> I should also note a cmake error in the Hera log about an incorrect older version being loaded. It did not cause any issues in building or results, and may have been due to me running another WM test, however if it reappears in future testing it would be a good idea to start tracking as a repo issue. I'll keep an eye out for it.

@FernandoAndrade-NOAA I didn't see any errors in my UPP building log under my account. The loaded cmake module is 3.23.1.

**Comment by WenMeng-NOAA** [comment]
This PR is ready for merging.

**Review by WenMeng-NOAA — APPROVED** [review]

**Pull request merged**